src/routes/(auth)/reset-password/+page.server.ts

```diff
@@ -0,0 +1,77 @@
+import { redirect, error } from '@sveltejs/kit';
+import type { PageServerLoad, Actions } from './$types';
+import { fail } from '@sveltejs/kit';
+import prisma from '@db';
+import { hashPassword } from '$lib/crypto';
+import { env } from '$env/dynamic/private';
+import { validatePassword } from '$lib/server/validate';
+
+export const load: PageServerLoad = async ({ url }) => {
+	const userId = url.searchParams.get('userId');
+	const token = url.searchParams.get('token');
+
+	if (!userId || !token) {
+		throw error(404, 'Not found');
+	}
+
+	const user = await prisma.user.findUnique({ where: { id: userId } });
+	const resetToken = await prisma.resetToken.findUnique({ where: { token: token } });
+
+	if (!user || !resetToken) {
+		throw error(404, 'Not found');
+	}
+
+	if (resetToken.expiresAt <= new Date()) {
+		throw error(404, 'Expired link');
+	}
+};
+
+export const actions: Actions = {
+	default: async ({ url, request }) => {
+		const userId = url.searchParams.get('userId');
+		if (!userId) throw error(404, 'Not found');
+
+		const user = await prisma.user.findUnique({ where: { id: userId } });
+		if (!user) throw error(404, 'Not found');
+
+		const data = await request.formData();
+		const newPassword = data.get('new-password');
+		const cnfPassword = data.get('confirm-password');
+		const errors: string[] = [];
+
+		if (!newPassword || !cnfPassword) {
+			errors.push('All fields are required');
+		}
+
+		if (newPassword !== cnfPassword) {
+			errors.push('Passwords need to match');
+		}
+
+		try {
+			if (newPassword) validatePassword(newPassword);
+		} catch (e: any) {
+			errors.push(e.message);
+		}
+
+		if (newPassword) {
+			const oldPasswordHash = user.password;
+			const newPasswordHash = await hashPassword(newPassword.toString(), env.SALT);
+			if (oldPasswordHash === newPasswordHash) {
+				errors.push('Cannot use existing password');
+			}
+		}
+
+		if (errors.length > 0) {
+			return fail(400, { success: false, errors });
+		}
+
+		if (newPassword && cnfPassword) {
+			const newPasswordHash = await hashPassword(newPassword.toString(), env.SALT);
+			await prisma.user.update({ where: { id: userId }, data: { password: newPasswordHash } });
+			await prisma.resetToken.delete({ where: { userId: userId } });
+			throw redirect(303, '/login');
```

Comment:
Instead of redirecting the user immediately after changing their password, I originally intended for the user to stay on the page with a confirmation message i.e. "Your password has been successfully updated." However, when I delete the reset token from the database, it forces a page reload, which is problematic since the token no longer will no longer exist and the route will be invalid. I attempted to suppress this page reload but was unable to find a way to do so. I'm open to suggestions or preferred ways to handle this use case.

Reply:
No "good" way around this issue, maybe we can implement toast messages later. For now, this is fine.